File: backend/controllers/auth_controller.py
from fastapi import APIRouter, HTTPException, Depends, status
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from models.schemas import LoginRequest, TokenResponse, TokenRefresh
from services.auth_service import auth_service
from motor.motor_asyncio import AsyncIOMotorDatabase
from typing import Optional
import logging

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/auth", tags=["Authentication"])
security = HTTPBearer()


async def get_current_user(
    credentials: HTTPAuthorizationCredentials = Depends(security),
    db: AsyncIOMotorDatabase = None
) -> dict:
    """Dependency to get current authenticated user"""
    token = credentials.credentials
    payload = auth_service.decode_token(token)
    
    if not payload:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid or expired token",
            headers={"WWW-Authenticate": "Bearer"}
        )
    
    if payload.get("type") != "access":
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid token type",
            headers={"WWW-Authenticate": "Bearer"}
        )
    
    return payload


def create_auth_routes(db: AsyncIOMotorDatabase):
    """Create authentication routes with database dependency"""
    
    @router.post("/login", response_model=TokenResponse)
    async def login(request: LoginRequest):
        """Authenticate user and return tokens"""
        # Find user
        user = await db.admin_users.find_one({"username": request.username}, {"_id": 0})
        
        if not user:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid credentials"
            )
        
        if not user.get("is_active", True):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Account is disabled"
            )
        
        if not auth_service.verify_password(request.password, user["password_hash"]):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid credentials"
            )
        
        # Create tokens
        tokens = auth_service.create_tokens(user["id"], user["username"])
        logger.info(f"User {request.username} logged in successfully")
        
        return TokenResponse(**tokens)
    
    @router.post("/refresh", response_model=TokenResponse)
    async def refresh_token(request: TokenRefresh):
        """Refresh access token using refresh token"""
        payload = auth_service.decode_token(request.refresh_token)
        
        if not payload:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid or expired refresh token"
            )
        
        if payload.get("type") != "refresh":
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid token type"
            )
        
        # Verify user still exists and is active
        user = await db.admin_users.find_one({"id": payload["sub"]}, {"_id": 0})
        if not user or not user.get("is_active", True):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="User not found or disabled"
            )
        
        # Create new tokens
        tokens = auth_service.create_tokens(user["id"], user["username"])
        logger.info(f"Token refreshed for user {user['username']}")
        
        return TokenResponse(**tokens)
    
    @router.get("/me")
    async def get_current_user_info(credentials: HTTPAuthorizationCredentials = Depends(security)):
        """Get current user information"""
        payload = auth_service.decode_token(credentials.credentials)
        
        if not payload or payload.get("type") != "access":
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid or expired token"
            )
        
        user = await db.admin_users.find_one({"id": payload["sub"]}, {"_id": 0, "password_hash": 0})
        if not user:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User not found"
            )
        
        return user
    
    return router

File: backend/controllers/AuthController.js
/**
 * Auth Controller
 * Single Responsibility: Handle authentication operations
 */

const bcrypt = require('bcryptjs');

class AuthController {
  constructor(collections, authMiddleware) {
    this.collections = collections;
    this.authMiddleware = authMiddleware;
  }

  // POST /api/auth/login
  login = async (req, res) => {
    try {
      const { username, password } = req.body;

      if (!username || !password) {
        return res.status(400).json({
          success: false,
          message: 'Username and password are required'
        });
      }

      // Find user
      const user = await this.collections.adminUsers.findOne(
        { username },
        { projection: { _id: 0 } }
      );

      if (!user) {
        return res.status(401).json({
          success: false,
          message: 'Invalid credentials'
        });
      }

      if (!user.is_active) {
        return res.status(401).json({
          success: false,
          message: 'Account is disabled'
        });
      }

      // Verify password
      const isValid = await bcrypt.compare(password, user.password_hash);
      if (!isValid) {
        return res.status(401).json({
          success: false,
          message: 'Invalid credentials'
        });
      }

      // Create tokens
      const tokens = this.authMiddleware.createTokens(user.id, user.username);

      console.log(`✅ User ${username} logged in`);

      res.json({
        success: true,
        ...tokens
      });
    } catch (error) {
      console.error('Login error:', error);
      res.status(500).json({
        success: false,
        message: 'Login failed'
      });
    }
  };

  // POST /api/auth/refresh
  refresh = async (req, res) => {
    try {
      const { refresh_token } = req.body;

      if (!refresh_token) {
        return res.status(400).json({
          success: false,
          message: 'Refresh token is required'
        });
      }

      const decoded = this.authMiddleware.verifyToken(refresh_token);

      if (!decoded) {
        return res.status(401).json({
          success: false,
          message: 'Invalid or expired refresh token'
        });
      }

      if (decoded.type !== 'refresh') {
        return res.status(401).json({
          success: false,
          message: 'Invalid token type'
        });
      }

      // Verify user still exists and is active
      const user = await this.collections.adminUsers.findOne(
        { id: decoded.sub },
        { projection: { _id: 0 } }
      );

      if (!user || !user.is_active) {
        return res.status(401).json({
          success: false,
          message: 'User not found or disabled'
        });
      }

      // Create new tokens
      const tokens = this.authMiddleware.createTokens(user.id, user.username);

      console.log(`✅ Token refreshed for user ${user.username}`);

      res.json({
        success: true,
        ...tokens
      });
    } catch (error) {
      console.error('Refresh error:', error);
      res.status(500).json({
        success: false,
        message: 'Token refresh failed'
      });
    }
  };

  // GET /api/auth/me
  getCurrentUser = async (req, res) => {
    try {
      const user = await this.collections.adminUsers.findOne(
        { id: req.user.id },
        { projection: { _id: 0, password_hash: 0 } }
      );

      if (!user) {
        return res.status(404).json({
          success: false,
          message: 'User not found'
        });
      }

      res.json({
        success: true,
        user
      });
    } catch (error) {
      console.error('Get current user error:', error);
      res.status(500).json({
        success: false,
        message: 'Failed to get user info'
      });
    }
  };
}

module.exports = AuthController;
